providers/digitalocean/digitalocean.go

```diff
@@ -0,0 +1,189 @@
+package digitalocean
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/Sirupsen/logrus"
+	api "github.com/digitalocean/godo"
+	"golang.org/x/oauth2"
+
+	"github.com/juju/ratelimit"
+	"github.com/rancher/external-dns/providers"
+	"github.com/rancher/external-dns/utils"
+)
+
+type DigitalOceanProvider struct {
+	client         *api.Client
+	rootDomainName string
+	TTL            int
+	limiter        *ratelimit.Bucket
+}
+
+func init() {
+	providers.RegisterProvider("digitalocean", &DigitalOceanProvider{})
+}
+
+type TokenSource struct {
+	AccessToken string
+}
+
+func (t *TokenSource) Token() (*oauth2.Token, error) {
+	token := &oauth2.Token{
+		AccessToken: t.AccessToken,
+	}
+	return token, nil
+}
+
+func (p *DigitalOceanProvider) Init(rootDomainName string) error {
+	var pat string
+	if pat = os.Getenv("DO_PAT"); len(pat) == 0 {
+		return fmt.Errorf("DO_PAT is not set")
+	}
+
+	tokenSource := &TokenSource{
+		AccessToken: pat,
+	}
+
+	oauthClient := oauth2.NewClient(oauth2.NoContext, tokenSource)
+	p.client = api.NewClient(oauthClient)
+
+	// DO's API is rate limited at 5000/hour.
+	doqps := (float64)(5000.0 / 3600.0)
+	p.limiter = ratelimit.NewBucketWithRate(doqps, 1)
```

Comment:
That isn't unreasonable.

Looking at the code, the Init routine requires 2 just to get off the ground while GetRecords requires 1 per page and AddRecord needs 1 per record to add.  So starting with say 50 or 100 would get medium-sized installations off the ground without waiting very long.

I also looked at the DO API documentation, and rate limiting is a *little* more complicated than a simple count of hits per time period.  I'll cut and paste the relevant docs here:

> 
> The way that a request expires and is removed from the current limit count is important to understand. Rather than counting all of the requests for an hour and resetting the RateLimit-Remaining value at the end of the hour, each request instead has its own timer.
> 
> This means that each request contributes toward the RateLimit-Remaining count for one complete hour after the request is made. When that request's timer runs out, it is no longer counted towards the request limit.
> 
> This has implications on the meaning of the RateLimit-Reset header as well. Because the entire rate limit is not reset at one time, the value of this header is set to the time when the oldest request will expire.

The ratelimit approach in this code is conservative enough that it should work under all but the highest of loads -- and those folks probably won't use DO. :-)